onnxruntime/python/onnxruntime_pybind_state.cc

```diff
@@ -517,8 +526,19 @@ void RegisterExecutionProviders(InferenceSession* sess, const std::vector<std::s
       RegisterExecutionProvider(sess, *onnxruntime::CreateExecutionProviderFactory_MIGraphX(0));
 #endif
     } else if (type == kCudaExecutionProvider) {
-#ifdef USE_CUDA
-
+#if (defined USE_CUDA && defined USE_DML)
+      // If both CUDA and DML are requested in this build, just register DML EP (below) -
+      // so no op here as ORT doesn't have semantics of dealing with nodes assigned to multiple
+      // device memory based GPU EPs. See comment and static_assert above.
+
+      // Adding logs to indicate that this is a no op as technically a user could try to
+      // register the CUDA EP given that it would be part of the list of available providers
+      // in a build that enables USE_CUDA
+      const logging::Logger& default_logger = logging::LoggingManager::DefaultLogger();
```

Comment:
Not sure this sort of hardcoded exception is the best way to address the test failure. It's technically valid to build with both as it's only an issue if you register both (which is a runtime choice) AND you have a model with nodes that will be assigned to the two different non-CPU EPs. If that happened in the real world the user could choose which of the EPs to register in order to avoid the problem.

Reply:
But in the case of python - building with both is the same as registering both the EPs (by default). A user would then have to look at the failure if they have a model assigned to both the EPs and then choose one. I think ideally we should not be building with both in the CI pipeline (which is often the first place people look while trying to build from source) and just restrict building with EP combinations that are not compatible together.

Reply:
> But in the case of python - building with both is the same as registering both the EPs (by default). A user would then have to look at the failure if they have a model assigned to both the EPs and then choose one (granted that this is rae but it doesn't look like good user experience). I think ideally we should not be building with both in the CI pipeline (which is often the first place people look while trying to build from source) and just restrict building with EP combinations that are not compatible together. For now, I ll just exclude the CUDA EP for dml build for this particular test .

I agree here. For python, there's implicit registration based on build options. We should avoid registering incompatible combinations there. it works for 99% of cases, but looks like as we enable more combination builds, we may have to revisit. i'd hate to have to enforce explicit registrations for python, as the defaults worked fine and was convenient for vast majority of users.

Reply:
Yeah. Ideally we should:

1) not build incompatible EPs together (like @Craigacp says). But this would mean having a new DML leg in the Windows GPU CI pipeline (bit like the nuget pipeline) and this is going to increase resource consumption.

2) not implicitly register incompatible EPs together (this was the goal of this PR). But even this had an element of ugliness as we had to allow CUDA and DML to be built together but not registered together.